README.md
<a href="http://www.jetbrains.com/idea/"><img
                src="http://resources.jetbrains.com/storage/products/intellij-idea/img/meta/intellij-idea_logo_300x300.png"
                width="100" height="100" border="0" alt="The best Java IDE" align="right"/></a>

## Retrotranslator


#### Contents

1.  [What is Retrotranslator?](#what)
2.  [What Java 5.0 features are supported?](#features)
3.  [How to use Retrotranslator from the command line?](#commandline)
4.  [How to use Retrotranslator from Apache Ant or Maven?](#ant)
5.  [How to use Retrotranslator from IntelliJ IDEA?](#idea)
6.  [How to produce a jar file compatible with Java 1.4?](#jarfile)
7.  [How to use Just-in-Time Retrotranslator?](#jit)
8.  [What API is supported on Java 1.4 and Java 1.3?](#supported)
9.  [How to write an extension?](#extension)
10.  [What are the limitations?](#limitations)
11.  [Alternative tools](#alternative)
12.  [Contact](#contact)
13.  [License](#license)

#### <a name="what">What is Retrotranslator?</a>

Retrotranslator is a tool that makes Java applications compatible with Java 1.4, Java 1.3 and other environments. It supports all Java 5.0 language features and a significant part of the Java 5.0 API on both J2SE 1.4 and J2SE 1.3. In other Java environments only the Java 5.0 features that don't depend on the new API are supported. Retrotranslator employs the [ASM](http://asm.objectweb.org/) bytecode manipulation framework to transform compiled Java classes and the [backport](http://backport-jsr166.sourceforge.net/) of the Java 5.0 concurrency utilities to emulate the Java 5.0 API.

#### <a name="features">What Java 5.0 features are supported?</a>

*   Generics
*   Annotations
*   Reflection on generics and annotations
*   Typesafe enums
*   Autoboxing/unboxing
*   Enhanced for loop
*   Varargs
*   Covariant return types
*   Formatted output
*   Static import
*   Concurrency utilities
*   Collections framework enhancements

#### <a name="commandline">How to use Retrotranslator from the command line?</a>

1.  [Download](http://sourceforge.net/project/showfiles.php?group_id=153566) and unzip the binary distribution file `Retrotranslator-_n.n.n_-bin.zip`, where _n.n.n_ is the latest Retrotranslator release number.
2.  Compile your classes with Java 5.0 or Java 6 and put them into some directory, e.g. `myclasses`.
3.  Go to the unzipped directory and execute:  
    `java -jar retrotranslator-transformer-_n.n.n_.jar -srcdir myclasses`  
    Use appropriate options to verify the result and for troubleshooting, e.g. -[verify](#option_verify), -[classpath](#option_classpath), -[advanced](#option_advanced), and -[smart](#option_smart).
4.  Put `retrotranslator-runtime-_n.n.n_.jar` and `backport-util-concurrent-_n.n_.jar` into the classpath of your application if you use the Java 5.0 API.
5.  Run or debug the application as usual on Java 1.4.

The command line syntax:  
`java -jar retrotranslator-transformer-_n.n.n_.jar <options>`  
or  
`java -cp retrotranslator-transformer-_n.n.n_.jar net.sf.retrotranslator.transformer.Retrotranslator <options>`

<table border="1" cellspacing="0" cellpadding="5">

<tbody>

<tr>

<th>Option</th>

<th>Description</th>

<th>Default</th>

</tr>

<tr>

<td nowrap="">

`-srcdir <path>`

</td>

<td>The directory with the files to process (can be specified several times).</td>

<td>-</td>

</tr>

<tr>

<td nowrap="">

`-srcjar <file>`

</td>

<td>The jar file with the files to process (can be specified several times).</td>

<td>-</td>

</tr>

<tr>

<td nowrap="">

`-destdir <path>`

</td>

<td>The directory to place processed files.</td>

<td nowrap="">The source.</td>

</tr>

<tr>

<td nowrap="">

`-destjar <file>`

</td>

<td>The jar file to place processed files.</td>

<td>The source.</td>

</tr>

<tr>

<td nowrap="">

`-srcmask <mask>`

</td>

<td>

The wildcard pattern specifying the files to transform rather than copy (classes or UTF-8 text files), e.g. `*.class;?*.tld`.

</td>

<td>

`*.class`

</td>

</tr>

<tr>

<td nowrap="">

<a name="option_target">`-target <version>`</a>

</td>

<td>

The version of the JVM where the translated application should be able to run (1.1, 1.2, 1.3, 1.4, or 1.5). The Java 5.0 API is supported only for the 1.4 and 1.3 targets, but user-defined [backport](#option_backport) classes can be used for other targets as well.

</td>

<td>

`1.4`

</td>

</tr>

<tr>

<td nowrap="">

<a name="option_classpath">`-classpath <path>`</a>

</td>

<td>

The dependencies of the translated classes including backport packages and jar files from the target JVM. For the 1.4 target the following files should be specified among others: `rt.jar`, `jce.jar`, and `jsse.jar` from Sun JRE 1.4; `retrotranslator-runtime-_n.n.n_.jar` and `backport-util-concurrent-_n.n_.jar`. For the 1.3 target the files are: `rt.jar` from Sun JRE 1.3, `retrotranslator-runtime13-_n.n.n_.jar` and `backport-util-concurrent-java12-_n.n_.jar`. This option can be omitted if the current Java environment matches the target one.

</td>

<td>The current classpath.</td>

</tr>

<tr>

<td nowrap="">

<a name="option_verify">`-verify`</a>

</td>

<td>

Asks the translator to examine translated bytecode for references to classes, methods, or fields that cannot be found in the [classpath](#option_classpath). Every use of Java 5.0 API will result in a warning message if the [classpath](#option_classpath) contains jar files from JRE 1.4.

</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

<a name="option_support">`-support <features>`</a>

</td>

<td>

Enables advanced features. The names specified should be separated with semicolons, e.g. `ThreadLocal.remove;BigDecimal.setScale`.

</td>

<td>-</td>

</tr>

<tr>

<td nowrap="">

<a name="option_advanced">`-advanced`</a>

</td>

<td>

Enables all advanced features at once, but it's recommended to enable only required features in order to avoid compatibility issues.

</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

<a name="option_smart">`-smart`</a>

</td>

<td>

Makes all backport classes inheritable provided that the [classpath](#option_classpath) correctly reflects the target environment. For example, the backport of the `Writer.append(String)` method can be used to translate the following expression: `new FileWriter("file.tmp").append("Hello")`.

</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

<a name="option_syncvolatile">`-syncvolatile`</a>

</td>

<td>

Makes access to volatile fields of Java 5.0 classes compliant with Java 5.0 memory model at the cost of decreased performance. Highly recommended to use.

</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

<a name="option_syncfinal">`-syncfinal`</a>

</td>

<td>

Makes access to final instance fields of Java 5.0 classes compliant with Java 5.0 memory model at the cost of decreased performance. Not recommended to use.

</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

<a name="option_backport">`-backport <names>`</a>

</td>

<td>

The [backport](#extension) names separated with semicolons, e.g. `com.myco.all;java.util:com.myco.ju;javax.net.SocketFactory:com.myco.jsse.Factory`. The corresponding backport classes must be present in the [classpath](#option_classpath).

</td>

<td>-</td>

</tr>

<tr>

<td nowrap="">

`-embed <package>`

</td>

<td>

The unique package name for a partial copy of `retrotranslator-runtime-_n.n.n_.jar` and `backport-util-concurrent-_n.n_.jar` to be put along with translated classes. This makes the translated application independent of other versions of Retrotranslator present in the classpath.

</td>

<td>-</td>

</tr>

<tr>

<td nowrap="">

`-lazy`

</td>

<td>

Asks the translator to transform and verify only the classes compiled with a target greater than the [current](#option_target) one.

</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

`-stripsign`

</td>

<td>Whether to remove signature (generics) information.</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

`-stripannot`

</td>

<td>Whether to remove annotations and related attributes.</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

`-verbose`

</td>

<td>Asks the translator for verbose output.</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

`-retainapi`

</td>

<td>

Asks the translator to modify classes for JVM compatibility but keep use of API unless overriden with the [backport](#option_backport) option. Any references introduced by a compiler remain unchanged, like the use of `java.lang.StringBuilder` for string concatenation or the implicit `valueOf` method calls for autoboxing.

</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

<a name="option_keepclasslit">`-keepclasslit`</a>

</td>

<td>

Prevents replacement of certain types like `java.lang.Iterable` or `java.util.Queue` with their base types in class literals.

</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

`-retainflags`

</td>

<td>Whether to keep Java 5.0 specific access modifiers.</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

`-uptodatecheck`

</td>

<td>Asks the translator to skip processing of files if the destination files are newer.</td>

<td>

`false`

</td>

</tr>

<tr>

<td nowrap="">

`-reflection <mode>`

</td>

<td>

Setting this option to `safe` makes the translator include additional metadata into classes, so reflection works even if the classes are unavailable as resources.

</td>

<td>

`normal`

</td>

</tr>

</tbody>

</table>

#### <a name="ant">How to use Retrotranslator from Apache Ant or Maven?</a>

The distribution contains an [Apache Ant](http://ant.apache.org/) task `net.sf.retrotranslator.transformer.RetrotranslatorTask`. Every [command line](#commandline) option can be set using the corresponding attribute. In addition the source files can be specified with nested `fileset`, `jarfileset`, and `dirset` elements and the [classpath](#option_classpath) can be set with nested `classpath` elements or the `classpathref` attribute. The source directories specified with `srcdir`, `dirset`, and the `dir` attribute of `fileset` should contain the root package of the classes. In case of warnings the build fails unless the value of the `failonwarning` attribute is set to `false`. The following script can be used to build one jar compatible with Java 1.4 and another one compatible with Java 1.3.

```xml
    <taskdef name="retrotranslator" classname="net.sf.retrotranslator.transformer.RetrotranslatorTask">
        <classpath>
            <fileset dir="../Retrotranslator-n.n.n-bin">
                <include name="retrotranslator-transformer-n.n.n.jar" />
                <include name="retrotranslator-runtime-n.n.n.jar" />
                <include name="backport-util-concurrent-n.n.jar" />
            </fileset>
        </classpath>
    </taskdef>

    <retrotranslator target="1.4" destjar="build/application14.jar"
                     smart="true" verify="true" failonwarning="false">
        <fileset dir="build/classes" includes="**/*.class" />
        <jarfileset dir="build/lib" includes="**/*.jar" />
        <classpath>
            <fileset dir="../j2sdk1.4.2_17/jre/lib" includes="**/*.jar"/>
            <fileset dir="../Retrotranslator-n.n.n-bin">
                <include name="retrotranslator-runtime-n.n.n.jar" />
                <include name="backport-util-concurrent-n.n.jar" />
            </fileset>
            <fileset dir="lib" includes="**/*.jar"/>
        </classpath>
    </retrotranslator>

    <retrotranslator target="1.3" destjar="build/application13.jar"
                     smart="true" verify="true" failonwarning="false">
        <fileset dir="build/classes" includes="**/*.class" />
        <jarfileset dir="build/lib" includes="**/*.jar" />
        <classpath>
            <fileset dir="../jdk1.3.1_20/jre/lib" includes="**/*.jar"/>
            <fileset dir="../Retrotranslator-n.n.n-bin">
                <include name="retrotranslator-runtime13-n.n.n.jar" />
                <include name="backport-util-concurrent-java12-n.n.jar" />
            </fileset>
            <fileset dir="lib" includes="**/*.jar"/>
        </classpath>
    </retrotranslator>
```

[Maven](http://maven.apache.org/) users can use the [Retrotranslator plugin](http://mojo.codehaus.org/retrotranslator-maven-plugin/) from the [Mojo Project](http://mojo.codehaus.org/).

#### <a name="idea">How to use Retrotranslator from IntelliJ IDEA?</a>

There is a [plugin](http://plugins.intellij.net/plugin/?id=145) to automatically translate and verify classes compiled by [IntelliJ IDEA](http://www.jetbrains.com/idea/), so you can develop in Java 5.0 but run and debug on Java 1.4.

#### <a name="jarfile">How to produce a jar file compatible with Java 1.4?</a>

If you have a `myapplication5.jar` file built with Java 5.0 you can use the following command to produce `myapplication4.jar`. The jar file will be compatible with Java 1.4 and independent of Retrotranslator because backport classes are added to the translated application with a different package name.

`java -jar retrotranslator-transformer-_n.n.n_.jar -srcjar myapplication5.jar -destjar myapplication4.jar -embed com.mycompany.internal`  

Also it is recommended to specify the [classpath](#option_classpath), [verify](#option_verify), and [smart](#option_smart) options. In case of verification failure try the [support](#option_support) option.

#### <a name="jit">How to use Just-in-Time Retrotranslator?</a>

In order to run a Java 5.0 application on Java 1.4 start it with JIT Retrotranslator using one of the following commands:

*   `java -cp retrotranslator-transformer-_n.n.n_.jar net.sf.retrotranslator.transformer.JITRetrotranslator <options> -jar <jarfile> [<args...>]`
*   `java -cp <classpath including Retrotranslator> net.sf.retrotranslator.transformer.JITRetrotranslator <options> <class> [<args...>]`

The options include [-support](#option_support), [-advanced](#option_advanced), [-smart](#option_smart), [-backport](#option_backport), [-syncvolatile](#option_syncvolatile), [-syncfinal](#option_syncfinal), and [-keepclasslit](#option_keepclasslit). When running on J2SE 5.0 JIT Retrotranslator simply calls the application, but on J2SE 1.4 it also translates classes compiled for Java 5.0 or later. However this capability depends on the current JVM and the application itself, so under certain conditions JIT Retrotranslator may be unable to translate either a jar file or classes from the classpath or both.

#### <a name="supported">What API is supported on Java 1.4 and Java 1.3?</a>

An asterisk designates all API members introduced in Java 5.0. Most classes and methods added in Java 1.4 and in Java 6.0 cannot be used on earlier VMs.

<table border="1" cellspacing="0" cellpadding="5">

<tbody>

<tr>

<th>Package</th>

<th>Class</th>

<th>Members</th>

<th>Compatibility notes</th>

</tr>

<tr>

<td rowspan="6">

`java.io`

</td>

<td>

`Closeable`<sup>[2](#2)</sup>

</td>

<td>* (all members introduced in Java 5.0)</td>

<td> </td>

</tr>

<tr>

<td>

`Flushable`<sup>[2](#2)</sup>

</td>

<td>*</td>

<td> </td>

</tr>

<tr>

<td>

`PrintStream`

</td>

<td>

\* (11 new methods and constructors)  
`PrintStream(File, String)`<sup>[5](#5)</sup>,  
`PrintStream(String, String)`<sup>[5](#5)</sup>

</td>

<td> </td>

</tr>

<tr>

<td>

`PrintWriter`

</td>

<td>

\* (11 new methods and constructors)

</td>

<td>

The `PrintWriter.format` and `PrintWriter.printf` methods always flush the output buffer.

</td>

</tr>

<tr>

<td>

`Reader`<sup>[5](#5)</sup>

</td>

<td>

\* (1 new method)

</td>

<td> </td>

</tr>

<tr>

<td>

`Writer`

</td>

<td>

\* (3 new methods)

</td>

<td> </td>

</tr>

<tr>

<td rowspan="32">

`java.lang`

</td>

<td>

`Appendable`<sup>[2](#2)</sup>

</td>

<td>*</td>

<td> </td>

</tr>

<tr>

<td>

`Boolean`

</td>

<td>

\* (2 new methods)  
`toString(boolean)`<sup>[6](#6)</sup>,  
`valueOf(boolean)`<sup>[6](#6)</sup>

</td>

<td> </td>

</tr>

<tr>

<td>

`Byte`

</td>

<td>

\* (1 new method)

</td>

<td> </td>

</tr>

<tr>

<td>

`Character`

</td>

<td>

\* (44 new methods)  
`getDirectionality(int)`<sup>[5](#5)</sup>,  
`isMirrored(int)`<sup>[5](#5)</sup>,  
`toString(char)`<sup>[6](#6)</sup>

</td>

<td>

New members of `Character.UnicodeBlock` are not supported. All supplementary code points are considered as unassigned.

</td>

</tr>

<tr>

<td>

`CharSequence`<sup>[6](#6)</sup>

</td>

<td>*</td>

<td> </td>

</tr>

<tr>

<td>

`Class`

</td>

<td>

\* (21 new methods)

</td>

<td>

Enable feature "`Class.forName`" to use backported classes as a fallback when calling this method<sup>[3](#3)</sup>.  
Enable features "`Class.getMethod`" and "`Class.getDeclaredMethod`" to use most of backported static methods as a fallback when calling these methods and for more uniform support of generics and covariant return types on different platforms<sup>[3](#3)</sup>.

</td>

</tr>

<tr>

<td>

`Deprecated`

</td>

<td>

\*

</td>

<td> </td>

</tr>

<tr>

<td>

`Double`

</td>

<td>

`valueOf(double)`

</td>

<td> </td>

</tr>

<tr>

<td>

`Enum`

</td>

<td>*</td>

<td> </td>

</tr>

<tr>

<td>

`Float`

</td>

<td>

`valueOf(float)`

</td>

<td> </td>

</tr>

<tr>

<td>

`IllegalArgumentException`

</td>

<td>

\* (2 new constructors)

</td>

<td> </td>

</tr>

<tr>

<td>

`IllegalStateException`

</td>

<td>

\* (2 new constructors)

</td>

<td> </td>

</tr>

<tr>

<td>

`Integer`

</td>

<td>

`valueOf(int),  
signum(int)`

</td>

<td> </td>

</tr>

<tr>

<td>

`Iterable`<sup>[2](#2)</sup>

</td>

<td>*</td>

<td> </td>

</tr>

<tr>

<td>

`Long`

</td>

<td>

`valueOf(long),  
signum(long)`

</td>

<td> </td>

</tr>

<tr>

<td>

`Math`

</td>

<td>

`cbrt(double), cosh(double),  
expm1(double), log10(double),  
log1p(double), signum(double),  
signum(float), sinh(double),  
tanh(double)`

</td>

<td>The quality of computation may be insufficient for applications of certain types.</td>

</tr>

<tr>

<td>

`Package`

</td>

<td>

\* (4 new methods)

</td>

<td>

Enable feature "`Retrotranslator.fixHyphen`" to change the name of `package-info` classes to `package$info`<sup>[3](#3)</sup>.

</td>

</tr>

<tr>

<td>

`Readable`<sup>[2](#2),[5](#5)</sup>

</td>

<td>*</td>

<td> </td>

</tr>

<tr>

<td>

`Short`

</td>

<td>

\* (2 new methods)

</td>

<td> </td>

</tr>

<tr>

<td>

`StackTraceElement`<sup>[6](#6)</sup>

</td>

<td>

\* (1 new constructor)

</td>

<td> </td>

</tr>

<tr>

<td>

`String`

</td>

<td>


changelog.txt
Retrotranslator Changelog
=================================

0.9.6 (2006-01-05)
---------------------------------
  - Dynamic discovery of backported classes.
  - Full support for generics reflection.
  - Improved java.lang.Enum support.
  - Support for all methods in java.lang.Class.
  - Support for BigDecimal(int), BigDecimal(long).
  - Iterable is distinct from Collection, but even Iterable arrays can hold Collections.
  - Patch 1392759 - build patch to create maven repository deployment units.

0.9.5 (2005-12-28)
---------------------------------
  - Support for EnumSet.
  - Support for methods in Arrays.
  - Partial support for generics reflection.
  - Patch 1392096 - easier integration to Maven 2.
  - Patch 1391894 - fix for bug 1391863.
  - Patch 1391842 - build patch for Mac OS X.
  - ASM 2.1 replaced with ASM 3.0 beta.

0.9.4 (2005-12-19)
---------------------------------
  - Support for methods in LinkedList: element, offer, peek, poll, remove.
  - Support for methods in Package: isAnnotationPresent, getAnnotation, getAnnotations, getDeclaredAnnotations.
  - A variable or field of type Queue is assignable from LinkedList.
  - A variable or field of type AnnotatedElement is assignable from Package, Class, Constructor, Field, or Method.
  - Bug fix: IOException when JIT Retrotranslator runs jar file with dependencies.

0.9.3 (2005-12-13)
---------------------------------
  - JIT Retrotranslator for Sun JVM 1.4 added. You may launch your application from
    net.sf.retrotranslator.transformer.JITRetrotranslator or simply
    call JITRetrotranslator.install() before loading Java 5.0 classes.
  - java.lang.Iterable is now replaced with java.util.Collection during translation.
  - Bug fix: Class.isAnnotation and Class.isEnum didn't work.

0.9.2 (2005-12-09)
---------------------------------
  - Support for methods: String.contains, String.contentEquals, String.replace, Matcher.quoteReplacement,
    Class.isAnnotation, Class.isEnum, Collections.checkedMap, Class.checkedSortedMap.
  - New option -stripsign to strip signature information used by Java 5.0 compiler.
  - Bug fix: the verifier didn't deal with abstract classes and arrays of primitives correctly.

0.9.1 (2005-12-08)
---------------------------------
  - A reference verifier to show warnings when references to unknown classes, methods, or fields are found.
  - An integrated Ant task.
  - Command line syntax changed:
     replaced: -source by -srcdir;
     added: -destdir, -verify, -classpath;
     removed: -target
  - Support for methods: Boolean.parseBoolean, Class.getEnumConstants, Class.cast, Class.asSubclass.
  - Support for new methods in Collections: checkedCollection, checkedSet, checkedSortedSet,
  checkedList, emptySet, emptyList, emptyMap, reverseOrder, frequency, disjoint, addAll.

0.9.0 (2005-11-28)
---------------------------------
Initial release


src/net/sf/retrotranslator/runtime/asm/Opcodes.java
/***
 * ASM: a very small and fast Java bytecode manipulation framework
 * Copyright (c) 2000-2005 INRIA, France Telecom
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the copyright holders nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 * THE POSSIBILITY OF SUCH DAMAGE.
 */
package net.sf.retrotranslator.runtime.asm;

/**
 * Defines the JVM opcodes, access flags and array type codes. This interface
 * does not define all the JVM opcodes because some opcodes are automatically
 * handled. For example, the xLOAD and xSTORE opcodes are automatically replaced
 * by xLOAD_n and xSTORE_n opcodes when possible. The xLOAD_n and xSTORE_n
 * opcodes are therefore not defined in this interface. Likewise for LDC,
 * automatically replaced by LDC_W or LDC2_W when necessary, WIDE, GOTO_W and
 * JSR_W.
 * 
 * @author Eric Bruneton
 * @author Eugene Kuleshov
 */
public interface Opcodes {

    // versions

    int V1_1 = 3 << 16 | 45;
    int V1_2 = 0 << 16 | 46;
    int V1_3 = 0 << 16 | 47;
    int V1_4 = 0 << 16 | 48;
    int V1_5 = 0 << 16 | 49;
    int V1_6 = 0 << 16 | 50;

    // access flags

    int ACC_PUBLIC = 0x0001; // class, field, method
    int ACC_PRIVATE = 0x0002; // class, field, method
    int ACC_PROTECTED = 0x0004; // class, field, method
    int ACC_STATIC = 0x0008; // field, method
    int ACC_FINAL = 0x0010; // class, field, method
    int ACC_SUPER = 0x0020; // class
    int ACC_SYNCHRONIZED = 0x0020; // method
    int ACC_VOLATILE = 0x0040; // field
    int ACC_BRIDGE = 0x0040; // method
    int ACC_VARARGS = 0x0080; // method
    int ACC_TRANSIENT = 0x0080; // field
    int ACC_NATIVE = 0x0100; // method
    int ACC_INTERFACE = 0x0200; // class
    int ACC_ABSTRACT = 0x0400; // class, method
    int ACC_STRICT = 0x0800; // method
    int ACC_SYNTHETIC = 0x1000; // class, field, method
    int ACC_ANNOTATION = 0x2000; // class
    int ACC_ENUM = 0x4000; // class(?) field inner

    // ASM specific pseudo access flags

    int ACC_DEPRECATED = 131072; // class, field, method

    // types for NEWARRAY

    int T_BOOLEAN = 4;
    int T_CHAR = 5;
    int T_FLOAT = 6;
    int T_DOUBLE = 7;
    int T_BYTE = 8;
    int T_SHORT = 9;
    int T_INT = 10;
    int T_LONG = 11;

    // opcodes // visit method (- = idem)

    int NOP = 0; // visitInsn
    int ACONST_NULL = 1; // -
    int ICONST_M1 = 2; // -
    int ICONST_0 = 3; // -
    int ICONST_1 = 4; // -
    int ICONST_2 = 5; // -
    int ICONST_3 = 6; // -
    int ICONST_4 = 7; // -
    int ICONST_5 = 8; // -
    int LCONST_0 = 9; // -
    int LCONST_1 = 10; // -
    int FCONST_0 = 11; // -
    int FCONST_1 = 12; // -
    int FCONST_2 = 13; // -
    int DCONST_0 = 14; // -
    int DCONST_1 = 15; // -
    int BIPUSH = 16; // visitIntInsn
    int SIPUSH = 17; // -
    int LDC = 18; // visitLdcInsn
    // int LDC_W = 19; // -
    // int LDC2_W = 20; // -
    int ILOAD = 21; // visitVarInsn
    int LLOAD = 22; // -
    int FLOAD = 23; // -
    int DLOAD = 24; // -
    int ALOAD = 25; // -
    // int ILOAD_0 = 26; // -
    // int ILOAD_1 = 27; // -
    // int ILOAD_2 = 28; // -
    // int ILOAD_3 = 29; // -
    // int LLOAD_0 = 30; // -
    // int LLOAD_1 = 31; // -
    // int LLOAD_2 = 32; // -
    // int LLOAD_3 = 33; // -
    // int FLOAD_0 = 34; // -
    // int FLOAD_1 = 35; // -
    // int FLOAD_2 = 36; // -
    // int FLOAD_3 = 37; // -
    // int DLOAD_0 = 38; // -
    // int DLOAD_1 = 39; // -
    // int DLOAD_2 = 40; // -
    // int DLOAD_3 = 41; // -
    // int ALOAD_0 = 42; // -
    // int ALOAD_1 = 43; // -
    // int ALOAD_2 = 44; // -
    // int ALOAD_3 = 45; // -
    int IALOAD = 46; // visitInsn
    int LALOAD = 47; // -
    int FALOAD = 48; // -
    int DALOAD = 49; // -
    int AALOAD = 50; // -
    int BALOAD = 51; // -
    int CALOAD = 52; // -
    int SALOAD = 53; // -
    int ISTORE = 54; // visitVarInsn
    int LSTORE = 55; // -
    int FSTORE = 56; // -
    int DSTORE = 57; // -
    int ASTORE = 58; // -
    // int ISTORE_0 = 59; // -
    // int ISTORE_1 = 60; // -
    // int ISTORE_2 = 61; // -
    // int ISTORE_3 = 62; // -
    // int LSTORE_0 = 63; // -
    // int LSTORE_1 = 64; // -
    // int LSTORE_2 = 65; // -
    // int LSTORE_3 = 66; // -
    // int FSTORE_0 = 67; // -
    // int FSTORE_1 = 68; // -
    // int FSTORE_2 = 69; // -
    // int FSTORE_3 = 70; // -
    // int DSTORE_0 = 71; // -
    // int DSTORE_1 = 72; // -
    // int DSTORE_2 = 73; // -
    // int DSTORE_3 = 74; // -
    // int ASTORE_0 = 75; // -
    // int ASTORE_1 = 76; // -
    // int ASTORE_2 = 77; // -
    // int ASTORE_3 = 78; // -
    int IASTORE = 79; // visitInsn
    int LASTORE = 80; // -
    int FASTORE = 81; // -
    int DASTORE = 82; // -
    int AASTORE = 83; // -
    int BASTORE = 84; // -
    int CASTORE = 85; // -
    int SASTORE = 86; // -
    int POP = 87; // -
    int POP2 = 88; // -
    int DUP = 89; // -
    int DUP_X1 = 90; // -
    int DUP_X2 = 91; // -
    int DUP2 = 92; // -
    int DUP2_X1 = 93; // -
    int DUP2_X2 = 94; // -
    int SWAP = 95; // -
    int IADD = 96; // -
    int LADD = 97; // -
    int FADD = 98; // -
    int DADD = 99; // -
    int ISUB = 100; // -
    int LSUB = 101; // -
    int FSUB = 102; // -
    int DSUB = 103; // -
    int IMUL = 104; // -
    int LMUL = 105; // -
    int FMUL = 106; // -
    int DMUL = 107; // -
    int IDIV = 108; // -
    int LDIV = 109; // -
    int FDIV = 110; // -
    int DDIV = 111; // -
    int IREM = 112; // -
    int LREM = 113; // -
    int FREM = 114; // -
    int DREM = 115; // -
    int INEG = 116; // -
    int LNEG = 117; // -
    int FNEG = 118; // -
    int DNEG = 119; // -
    int ISHL = 120; // -
    int LSHL = 121; // -
    int ISHR = 122; // -
    int LSHR = 123; // -
    int IUSHR = 124; // -
    int LUSHR = 125; // -
    int IAND = 126; // -
    int LAND = 127; // -
    int IOR = 128; // -
    int LOR = 129; // -
    int IXOR = 130; // -
    int LXOR = 131; // -
    int IINC = 132; // visitIincInsn
    int I2L = 133; // visitInsn
    int I2F = 134; // -
    int I2D = 135; // -
    int L2I = 136; // -
    int L2F = 137; // -
    int L2D = 138; // -
    int F2I = 139; // -
    int F2L = 140; // -
    int F2D = 141; // -
    int D2I = 142; // -
    int D2L = 143; // -
    int D2F = 144; // -
    int I2B = 145; // -
    int I2C = 146; // -
    int I2S = 147; // -
    int LCMP = 148; // -
    int FCMPL = 149; // -
    int FCMPG = 150; // -
    int DCMPL = 151; // -
    int DCMPG = 152; // -
    int IFEQ = 153; // visitJumpInsn
    int IFNE = 154; // -
    int IFLT = 155; // -
    int IFGE = 156; // -
    int IFGT = 157; // -
    int IFLE = 158; // -
    int IF_ICMPEQ = 159; // -
    int IF_ICMPNE = 160; // -
    int IF_ICMPLT = 161; // -
    int IF_ICMPGE = 162; // -
    int IF_ICMPGT = 163; // -
    int IF_ICMPLE = 164; // -
    int IF_ACMPEQ = 165; // -
    int IF_ACMPNE = 166; // -
    int GOTO = 167; // -
    int JSR = 168; // -
    int RET = 169; // visitVarInsn
    int TABLESWITCH = 170; // visiTableSwitchInsn
    int LOOKUPSWITCH = 171; // visitLookupSwitch
    int IRETURN = 172; // visitInsn
    int LRETURN = 173; // -
    int FRETURN = 174; // -
    int DRETURN = 175; // -
    int ARETURN = 176; // -
    int RETURN = 177; // -
    int GETSTATIC = 178; // visitFieldInsn
    int PUTSTATIC = 179; // -
    int GETFIELD = 180; // -
    int PUTFIELD = 181; // -
    int INVOKEVIRTUAL = 182; // visitMethodInsn
    int INVOKESPECIAL = 183; // -
    int INVOKESTATIC = 184; // -
    int INVOKEINTERFACE = 185; // -
    // int UNUSED = 186; // NOT VISITED
    int NEW = 187; // visitTypeInsn
    int NEWARRAY = 188; // visitIntInsn
    int ANEWARRAY = 189; // visitTypeInsn
    int ARRAYLENGTH = 190; // visitInsn
    int ATHROW = 191; // -
    int CHECKCAST = 192; // visitTypeInsn
    int INSTANCEOF = 193; // -
    int MONITORENTER = 194; // visitInsn
    int MONITOREXIT = 195; // -
    // int WIDE = 196; // NOT VISITED
    int MULTIANEWARRAY = 197; // visitMultiANewArrayInsn
    int IFNULL = 198; // visitJumpInsn
    int IFNONNULL = 199; // -
    // int GOTO_W = 200; // -
    // int JSR_W = 201; // -
}
